### PR TITLE
make DCTAdam acknowledge subsample

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #2386.

There are some changes that I haven't optimized such as separating the case `mask` is full index. I think it would be fine because the optimizer step is cheap. Please let me know if we need to do profiling those changes.

TODO:
- [x] add tests
- [x] incorporate reviewers' ideas

### CODE REVIEW

File: pyro/optim/dct_adam.py [hunk 1]
```diff
@@ -33,6 +31,20 @@ def _transform_inverse(x, dim, duration):
     return idct(x[right], dim)[left].add_(x[left])
 
 
+def _get_mask(x, indices):
+    # create a bollean mask from subsample indices
```
Comment (martinjankowiak) [inline pyro/optim/dct_adam.py:35]: typo: boolean

File: pyro/optim/dct_adam.py [hunk 2]
```diff
@@ -33,6 +31,20 @@ def _transform_inverse(x, dim, duration):
     return idct(x[right], dim)[left].add_(x[left])
 
 
+def _get_mask(x, indices):
```
Comment (martinjankowiak) [inline pyro/optim/dct_adam.py:34]: i'm having a bit of trouble understanding this helper. can you please either add more comments or add a test?

File: pyro/optim/dct_adam.py [hunk 3]
```diff
@@ -86,31 +101,36 @@ def step(self, closure=None):
...
-                    p.data.addcdiv_(-step_size, exp_avg, denom)
+                    step = exp_avg[mask] / denom
+                    p.data[mask] += (step.mul_(-step_size))
                 else:
-                    step = _transform_inverse(exp_avg / denom, time_dim, duration)
-                    p.data.add_(step.mul_(-step_size))
+                    step = _transform_inverse(exp_avg[mask] / denom, time_dim, duration)
+                    p.data[mask] += (step.mul_(-step_size))
+
+                if hasattr(p, "_pyro_subsample"):
```
Comment (martinjankowiak) [inline pyro/optim/dct_adam.py:133]: why is this necessary?

File: pyro/poutine/subsample_messenger.py [hunk 4]
```diff
@@ -149,4 +149,17 @@ def _postprocess_message(self, msg):
                             .format(self.name, self.size, self.dim, statement, shape))
                     # Subsample parameters with known batch semantics.
                     if self.subsample_size < self.size:
-                        msg["value"] = msg["value"].index_select(dim, self._indices)
+                        value = msg["value"]
```
Comment (martinjankowiak) [inline pyro/poutine/subsample_messenger.py:152]: why was the following logic necessary?

File: pyro/optim/dct_adam.py [hunk 5]
```diff
@@ -33,6 +31,20 @@ def _transform_inverse(x, dim, duration):
     return idct(x[right], dim)[left].add_(x[left])
 
 
+def _get_mask(x, indices):
+    # create a bollean mask from subsample indices
+    mask = x.new_zeros(x.shape, dtype=torch.long, device=x.device)
+    if len(indices) > 0:
+        size = x.dim()
+        idx = []
+        for dim in range(size):
```
Comment (fritzo) [inline pyro/optim/dct_adam.py:40]: nit: `for dim in range(-x.dim(), 0):` would avoid the need for `neg_dim`

File: pyro/optim/dct_adam.py (hunk 3 above)
Comment (fehiepsi) [reply pyro/optim/dct_adam.py:133]: This is not necessarily required. I just want to make sure those extra attributes are cleared if, for some reason, users want to run SVI with subsampling at the early training phase and run SVI without subsampling at later steps. This might be not a good place to put that logic. Do you have other suggestions?

File: pyro/poutine/subsample_messenger.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/poutine/subsample_messenger.py:152]: The logic is used to add `_pyro_subsample` attribute to the parameters which need to be optimized. I think that checking `grad=0` might affect non-subsampled parameters (e.g. when `loss = 2 * param +...` and `param` has the initial value is `0`).

File: pyro/optim/dct_adam.py (hunk 5 above)
Comment (fehiepsi) [reply pyro/optim/dct_adam.py:40]: Thanks!

File: pyro/optim/dct_adam.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/optim/dct_adam.py:35]: thanks!

File: pyro/optim/dct_adam.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/optim/dct_adam.py:36]: I just add some comments to explain the logic. :)

**Comment by fehiepsi** [comment]
Thank @fritzo and @martinjankowiak for reviewing and helpful suggestions at #2386! I have made corresponding changes. There is one minor issue is whether we clear "pyro_subsample" attribute at `DCTAdam.step` or someplace else. It is not important IMO so we might just skip it.

File: pyro/optim/dct_adam.py [hunk 6]
```diff
@@ -73,44 +91,95 @@ def step(self, closure=None):
...
+            # Exponential moving average of squared gradient values
+            state['exp_avg_sq'] = torch.zeros_like(p)
+
+        exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+        beta1, beta2 = group['betas']
+
+        state['step'][mask] += 1
+
+        # Decay the first and second moment running average coefficient
+        exp_avg.masked_scatter_(mask, exp_avg[mask].mul_(beta1).add_(1 - beta1, grad))
```
Comment (fehiepsi) [inline pyro/optim/dct_adam.py:172]: @martinjankowiak using `masked_scatter_` is faster than `exp_avg[mask] = ...`

**Comment by fehiepsi** [comment]
Hmm, looks like Travis is not working today.

File: pyro/poutine/subsample_messenger.py (hunk 4 above)
Comment (fritzo) [reply pyro/poutine/subsample_messenger.py:152]: It seems cleanest to me to use purely mask-based logic if possible, and assume it "almost surely" works. Can you think of a counterexample other than the first step example above?

**Comment by fehiepsi** [comment]
> why 54 -> 78?

It is because of those `x[mask]`. I have removed duplicated ones and seen the speed much improved. (another source is `beta1 ** state_step`; before this PR, it is scalar op; now it is tensor op)

File: pyro/poutine/subsample_messenger.py (hunk 4 above)
Comment (fehiepsi) [reply pyro/poutine/subsample_messenger.py:152]: I can't come up with a good example which shows the shortcomings of using `mask = (grad != 0)`: one example I have in mind is when a parameter moves to a local maximum area, which can have nonzero measure due to precision issue - to get out of it, we need to update the parameter by using grad information from past steps.

File: pyro/optim/dct_adam.py [hunk 7]
```diff
@@ -73,44 +94,98 @@ def step(self, closure=None):
...
+            state['step'] = torch.zeros_like(p)
+            # Exponential moving average of gradient values
+            state['exp_avg'] = torch.zeros_like(p)
+            # Exponential moving average of squared gradient values
+            state['exp_avg_sq'] = torch.zeros_like(p)
+
+        beta1, beta2 = group['betas']
+
+        state_step = state['step'].masked_select(mask).add_(1)
+        state['step'].masked_scatter_(mask, state_step)
```
Comment (fritzo) [inline pyro/optim/dct_adam.py:172]: Interesting pattern! Consider abstracting this out as a `masked_add_()` helper in `pyro.distributions.util`.

File: pyro/optim/dct_adam.py [hunk 8]
```diff
@@ -51,10 +69,13 @@ class DCTAdam(Optimizer):
         numerical stability (default: 1e-8)
     :param float clip_norm: magnitude of norm to which gradients are clipped (default: 10.0)
     :param float lrd: rate at which learning rate decays (default: 1.0)
+    :param bool aware_subsample: whether to update gradient statistics only for
```
Comment (fritzo) [inline pyro/optim/dct_adam.py:72]: nit: could we name this arg `subsample_aware`?

File: pyro/optim/dct_adam.py (hunk 7 above)
Comment (fehiepsi) [reply pyro/optim/dct_adam.py:172]: I would like to skip this abstraction. At other places, we not only use `add` but also `mul` or `addcmul`.
